diff -uNr SDL-1.2.15/src/cdrom/linux/SDL_syscdrom.c SDL-1.2.15.mod/src/cdrom/linux/SDL_syscdrom.c
--- SDL-1.2.15/src/cdrom/linux/SDL_syscdrom.c	2012-01-19 08:30:06.000000000 +0200
+++ SDL-1.2.15.mod/src/cdrom/linux/SDL_syscdrom.c	2017-11-26 00:28:55.341353617 +0200
@@ -68,12 +68,12 @@
 #ifdef MNTTAB
 #define _PATH_MNTTAB	MNTTAB
 #else
-#define _PATH_MNTTAB	"/etc/fstab"
+#define _PATH_MNTTAB	"/data/data/com.termux/files/usr/etc/fstab"
 #endif
 #endif /* !_PATH_MNTTAB */
 
 #ifndef _PATH_MOUNTED
-#define _PATH_MOUNTED	"/etc/mtab"
+#define _PATH_MOUNTED	"/data/data/com.termux/files/usr/etc/mtab"
 #endif /* !_PATH_MOUNTED */
 
 #ifndef MNTTYPE_CDROM
diff -uNr SDL-1.2.15/src/joystick/linux/SDL_sysjoystick.c SDL-1.2.15.mod/src/joystick/linux/SDL_sysjoystick.c
--- SDL-1.2.15/src/joystick/linux/SDL_sysjoystick.c	2012-01-19 08:30:06.000000000 +0200
+++ SDL-1.2.15.mod/src/joystick/linux/SDL_sysjoystick.c	2017-11-26 00:27:30.421353677 +0200
@@ -1106,6 +1106,9 @@
 				}
 				break;
 			    case EV_ABS:
+				if (code > ABS_MISC) {
+					break;
+				}
 				switch (code) {
 				    case ABS_HAT0X:
 				    case ABS_HAT0Y:
diff -uNr SDL-1.2.15/src/video/SDL_gamma.c SDL-1.2.15.mod/src/video/SDL_gamma.c
--- SDL-1.2.15/src/video/SDL_gamma.c	2012-01-19 08:30:06.000000000 +0200
+++ SDL-1.2.15.mod/src/video/SDL_gamma.c	2017-11-26 00:27:46.551353666 +0200
@@ -92,22 +92,11 @@
 
 int SDL_SetGamma(float red, float green, float blue)
 {
-	int succeeded;
+	int succeeded = -1;
 	SDL_VideoDevice *video = current_video;
 	SDL_VideoDevice *this  = current_video;	
 
-	succeeded = -1;
-	/* Prefer using SetGammaRamp(), as it's more flexible */
-	{
-		Uint16 ramp[3][256];
-
-		CalculateGammaRamp(red, ramp[0]);
-		CalculateGammaRamp(green, ramp[1]);
-		CalculateGammaRamp(blue, ramp[2]);
-		succeeded = SDL_SetGammaRamp(ramp[0], ramp[1], ramp[2]);
-	}
-	if ( (succeeded < 0) && video->SetGamma ) {
-		SDL_ClearError();
+	if ( video->SetGamma ) {
 		succeeded = video->SetGamma(this, red, green, blue);
 	}
 	return succeeded;
diff -uNr SDL-1.2.15/src/video/SDL_yuv_sw.c SDL-1.2.15.mod/src/video/SDL_yuv_sw.c
--- SDL-1.2.15/src/video/SDL_yuv_sw.c	2012-01-19 08:30:06.000000000 +0200
+++ SDL-1.2.15.mod/src/video/SDL_yuv_sw.c	2017-11-26 00:28:05.331353652 +0200
@@ -89,6 +89,8 @@
 #include "SDL_yuvfuncs.h"
 #include "SDL_yuv_sw_c.h"
 
+#undef __OPTIMIZE__
+
 /* The functions used to manipulate software video overlays */
 static struct private_yuvhwfuncs sw_yuvfuncs = {
 	SDL_LockYUV_SW,
diff -uNr SDL-1.2.15/src/video/fbcon/SDL_fbvideo.c SDL-1.2.15.mod/src/video/fbcon/SDL_fbvideo.c
--- SDL-1.2.15/src/video/fbcon/SDL_fbvideo.c	2012-01-19 08:30:06.000000000 +0200
+++ SDL-1.2.15.mod/src/video/fbcon/SDL_fbvideo.c	2017-11-26 00:29:03.621353611 +0200
@@ -280,7 +280,7 @@
 	FB_Available, FB_CreateDevice
 };
 
-#define FB_MODES_DB	"/etc/fb.modes"
+#define FB_MODES_DB	"/data/data/com.termux/files/usr/etc/fb.modes"
 
 static int read_fbmodes_line(FILE*f, char* line, int length)
 {
diff -uNr SDL-1.2.15/src/video/ipod/SDL_ipodvideo.c SDL-1.2.15.mod/src/video/ipod/SDL_ipodvideo.c
--- SDL-1.2.15/src/video/ipod/SDL_ipodvideo.c	2012-01-19 08:30:06.000000000 +0200
+++ SDL-1.2.15.mod/src/video/ipod/SDL_ipodvideo.c	2017-11-26 00:29:07.331353608 +0200
@@ -170,7 +170,7 @@
 	static const char * const vcs[] = { "/dev/vc/%d", "/dev/tty%d", 0 };
 	int i, tty0_fd;
 
-	dbgout = fdopen (open ("/etc/sdlpod.log", O_WRONLY | O_SYNC | O_APPEND), "a");
+	dbgout = fdopen (open ("/data/data/com.termux/files/usr/etc/sdlpod.log", O_WRONLY | O_SYNC | O_APPEND), "a");
 	if (dbgout) {
 	    setbuf (dbgout, 0);
 	    fprintf (dbgout, "--> Started SDL <--\n");
diff -uNr SDL-1.2.15/src/video/photon/SDL_ph_gl.h SDL-1.2.15.mod/src/video/photon/SDL_ph_gl.h
--- SDL-1.2.15/src/video/photon/SDL_ph_gl.h	2012-01-19 08:30:06.000000000 +0200
+++ SDL-1.2.15.mod/src/video/photon/SDL_ph_gl.h	2017-11-26 00:30:06.831353565 +0200
@@ -26,7 +26,7 @@
 
 #include "SDL_ph_video.h"
 
-#define DEFAULT_OPENGL "/usr/lib/libGL.so"
+#define DEFAULT_OPENGL "/data/data/com.termux/files/usr/lib/libGL.so"
 
 #if SDL_VIDEO_OPENGL
     void  ph_GL_SwapBuffers(_THIS);
diff -uNr SDL-1.2.15/src/video/x11/SDL_x11events.c SDL-1.2.15.mod/src/video/x11/SDL_x11events.c
--- SDL-1.2.15/src/video/x11/SDL_x11events.c	2012-01-19 08:30:06.000000000 +0200
+++ SDL-1.2.15.mod/src/video/x11/SDL_x11events.c	2017-11-26 00:28:11.471353648 +0200
@@ -57,12 +57,6 @@
 static SDLKey MISC_keymap[256];
 SDLKey X11_TranslateKeycode(Display *display, KeyCode kc);
 
-/*
- Pending resize target for ConfigureNotify (so outdated events don't
- cause inappropriate resize events)
-*/
-int X11_PendingConfigureNotifyWidth = -1;
-int X11_PendingConfigureNotifyHeight = -1;
 
 #ifdef X_HAVE_UTF8_STRING
 Uint32 Utf8ToUcs4(const Uint8 *utf8)
@@ -395,6 +389,8 @@
 {
 	int posted;
 	XEvent xevent;
+	int orig_event_type;
+	KeyCode orig_keycode;
 
 	SDL_memset(&xevent, '\0', sizeof (XEvent));  /* valgrind fix. --ryan. */
 	XNextEvent(SDL_Display, &xevent);
@@ -410,9 +406,29 @@
 #ifdef X_HAVE_UTF8_STRING
 	/* If we are translating with IM, we need to pass all events
 	   to XFilterEvent, and discard those filtered events immediately.  */
+	orig_event_type = xevent.type;
+	if (orig_event_type == KeyPress || orig_event_type == KeyRelease) {
+	     orig_keycode = xevent.xkey.keycode;
+	} else {
+	     orig_keycode = 0;
+	}
 	if ( SDL_TranslateUNICODE
 	     && SDL_IM != NULL
 	     && XFilterEvent(&xevent, None) ) {
+	        if (orig_keycode) {
+	            SDL_keysym keysym;
+	            static XComposeStatus state;
+	            char keybuf[32];
+
+	            keysym.scancode = xevent.xkey.keycode;
+	            keysym.sym = X11_TranslateKeycode(SDL_Display, xevent.xkey.keycode);
+	            keysym.mod = KMOD_NONE;
+	            keysym.unicode = 0;
+	            if (orig_event_type == KeyPress && XLookupString(&xevent.xkey, keybuf, sizeof(keybuf), NULL, &state))
+	                keysym.unicode = (Uint8)keybuf[0];
+
+	            SDL_PrivateKeyboard(orig_event_type == KeyPress ? SDL_PRESSED : SDL_RELEASED, &keysym);
+	        }
 		return 0;
 	}
 #endif
@@ -429,12 +445,15 @@
 if ( xevent.xcrossing.mode == NotifyUngrab )
 printf("Mode: NotifyUngrab\n");
 #endif
-		if ( this->input_grab == SDL_GRAB_OFF ) {
-			posted = SDL_PrivateAppActive(1, SDL_APPMOUSEFOCUS);
+		if ( (xevent.xcrossing.mode != NotifyGrab) &&
+		     (xevent.xcrossing.mode != NotifyUngrab) ) {
+			if ( this->input_grab == SDL_GRAB_OFF ) {
+				posted = SDL_PrivateAppActive(1, SDL_APPMOUSEFOCUS);
+			}
+			posted = SDL_PrivateMouseMotion(0, 0,
+					xevent.xcrossing.x,
+					xevent.xcrossing.y);
 		}
-		posted = SDL_PrivateMouseMotion(0, 0,
-				xevent.xcrossing.x,
-				xevent.xcrossing.y);
 	    }
 	    break;
 
@@ -825,16 +844,6 @@
 #ifdef DEBUG_XEVENTS
 printf("ConfigureNotify! (resize: %dx%d)\n", xevent.xconfigure.width, xevent.xconfigure.height);
 #endif
-		if ((X11_PendingConfigureNotifyWidth != -1) &&
-		    (X11_PendingConfigureNotifyHeight != -1)) {
-		    if ((xevent.xconfigure.width != X11_PendingConfigureNotifyWidth) &&
-			(xevent.xconfigure.height != X11_PendingConfigureNotifyHeight)) {
-			    /* Event is from before the resize, so ignore. */
-			    break;
-		    }
-		    X11_PendingConfigureNotifyWidth = -1;
-		    X11_PendingConfigureNotifyHeight = -1;
-		}
 		if ( SDL_VideoSurface ) {
 		    if ((xevent.xconfigure.width != SDL_VideoSurface->w) ||
 		        (xevent.xconfigure.height != SDL_VideoSurface->h)) {
@@ -1246,8 +1255,11 @@
  * sequences (dead accents, compose key sequences) will not work since the
  * state has been irrevocably lost.
  */
+extern DECLSPEC Uint16 SDLCALL X11_KeyToUnicode(SDLKey, SDLMod);
+
 Uint16 X11_KeyToUnicode(SDLKey keysym, SDLMod modifiers)
 {
+	static int warning = 0;
 	struct SDL_VideoDevice *this = current_video;
 	char keybuf[32];
 	int i;
@@ -1255,6 +1267,12 @@
 	XKeyEvent xkey;
 	Uint16 unicode;
 
+	if ( warning ) {
+		warning = 0;
+		fprintf(stderr, "WARNING: Application is using X11_KeyToUnicode().\n");
+		fprintf(stderr, "This is not an official SDL function, please report this as a bug.\n");
+	}
+
 	if ( !this || !SDL_Display ) {
 		return 0;
 	}
diff -uNr SDL-1.2.15/src/video/x11/SDL_x11events_c.h SDL-1.2.15.mod/src/video/x11/SDL_x11events_c.h
--- SDL-1.2.15/src/video/x11/SDL_x11events_c.h	2012-01-19 08:30:06.000000000 +0200
+++ SDL-1.2.15.mod/src/video/x11/SDL_x11events_c.h	2017-11-26 00:27:59.091353657 +0200
@@ -27,8 +27,3 @@
 extern void X11_InitOSKeymap(_THIS);
 extern void X11_PumpEvents(_THIS);
 extern void X11_SetKeyboardState(Display *display, const char *key_vec);
-
-/* Variables to be exported */
-extern int X11_PendingConfigureNotifyWidth;
-extern int X11_PendingConfigureNotifyHeight;
-
diff -uNr SDL-1.2.15/src/video/x11/SDL_x11gl.c SDL-1.2.15.mod/src/video/x11/SDL_x11gl.c
--- SDL-1.2.15/src/video/x11/SDL_x11gl.c	2012-01-19 08:30:06.000000000 +0200
+++ SDL-1.2.15.mod/src/video/x11/SDL_x11gl.c	2017-11-26 00:31:13.221353518 +0200
@@ -30,7 +30,7 @@
 /* IRIX doesn't have a GL library versioning system */
 #define DEFAULT_OPENGL	"libGL.so"
 #elif defined(__MACOSX__)
-#define DEFAULT_OPENGL	"/usr/X11R6/lib/libGL.1.dylib"
+#define DEFAULT_OPENGL	"/data/data/com.termux/files/usr/X11R6/lib/libGL.1.dylib"
 #elif defined(__QNXNTO__)
 #define DEFAULT_OPENGL	"libGL.so.3"
 #elif defined(__OpenBSD__)
diff -uNr SDL-1.2.15/src/video/x11/SDL_x11modes.c SDL-1.2.15.mod/src/video/x11/SDL_x11modes.c
--- SDL-1.2.15/src/video/x11/SDL_x11modes.c	2012-01-19 08:30:06.000000000 +0200
+++ SDL-1.2.15.mod/src/video/x11/SDL_x11modes.c	2017-11-26 00:30:52.121353533 +0200
@@ -458,7 +458,7 @@
     if ( SDL_strcmp(ServerVendor(SDL_Display), "Metro Link Incorporated") == 0 ) {
         FILE *metro_fp;
 
-        metro_fp = fopen("/usr/X11R6/lib/X11/Metro/.version", "r");
+        metro_fp = fopen("/data/data/com.termux/files/usr/X11R6/lib/X11/Metro/.version", "r");
         if ( metro_fp != NULL ) {
             int major, minor, patch, version, scannum;
             major = 0; minor = 0; patch = 0;
diff -uNr SDL-1.2.15/src/video/x11/SDL_x11sym.h SDL-1.2.15.mod/src/video/x11/SDL_x11sym.h
--- SDL-1.2.15/src/video/x11/SDL_x11sym.h	2012-01-19 08:30:06.000000000 +0200
+++ SDL-1.2.15.mod/src/video/x11/SDL_x11sym.h	2017-11-26 00:27:23.321353682 +0200
@@ -165,7 +165,7 @@
  */
 #ifdef LONG64
 SDL_X11_MODULE(IO_32BIT)
-SDL_X11_SYM(int,_XData32,(Display *dpy,register long *data,unsigned len),(dpy,data,len),return)
+SDL_X11_SYM(int,_XData32,(Display *dpy,register _Xconst long *data,unsigned len),(dpy,data,len),return)
 SDL_X11_SYM(void,_XRead32,(Display *dpy,register long *data,long len),(dpy,data,len),)
 #endif
 
diff -uNr SDL-1.2.15/src/video/x11/SDL_x11video.c SDL-1.2.15.mod/src/video/x11/SDL_x11video.c
--- SDL-1.2.15/src/video/x11/SDL_x11video.c	2012-01-19 08:30:06.000000000 +0200
+++ SDL-1.2.15.mod/src/video/x11/SDL_x11video.c	2017-11-26 00:27:59.101353657 +0200
@@ -1088,7 +1088,7 @@
 		}
 	}
 
-#if 0 /* This is an experiment - are the graphics faster now? - nope. */
+#if 1 /* This is an experiment - are the graphics faster now? - nope. */
 	if ( SDL_getenv("SDL_VIDEO_X11_BACKINGSTORE") )
 #endif
 	/* Cache the window in the server, when possible */
@@ -1177,8 +1177,6 @@
 			current = NULL;
 			goto done;
 		}
-		X11_PendingConfigureNotifyWidth = width;
-		X11_PendingConfigureNotifyHeight = height;
 	} else {
 		if (X11_CreateWindow(this,current,width,height,bpp,flags) < 0) {
 			current = NULL;
diff -uNr SDL-1.2.15/src/video/x11/SDL_x11wm.c SDL-1.2.15.mod/src/video/x11/SDL_x11wm.c
--- SDL-1.2.15/src/video/x11/SDL_x11wm.c	2012-01-19 08:30:06.000000000 +0200
+++ SDL-1.2.15.mod/src/video/x11/SDL_x11wm.c	2017-11-26 00:27:08.341353693 +0200
@@ -352,13 +352,14 @@
 			result = XGrabPointer(SDL_Display, SDL_Window, True, 0,
 						GrabModeAsync, GrabModeAsync,
 						SDL_Window, None, CurrentTime);
-			if ( result == GrabSuccess ) {
+			if ( result == GrabSuccess || result == GrabNotViewable ) {
 				break;
 			}
 			SDL_Delay(100);
 		}
 		if ( result != GrabSuccess ) {
 			/* Uh, oh, what do we do here? */ ;
+			return(SDL_GRAB_OFF);
 		}
 		/* Now grab the keyboard */
 		XGrabKeyboard(SDL_Display, WMwindow, True,
